src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -2850,3 +2851,44 @@ InstantiatePartitionQual(Node *node, void *context)
 
 	return expression_tree_mutator(node, InstantiatePartitionQual, context);
 }
+
+
+/*
+ * ErrorIfRouterSelectQueryNotSupported checks if the query contains unsupported features,
+ * and errors out if it does. Currently we only error out if query contains common table
+ * expressions where CTE part of the query is something other than SELECT.
+ */
+static void
+ErrorIfRouterSelectQueryNotSupported(Query *queryTree)
+{
+	ListCell *cteCell;
+
+	Assert(queryTree->commandType == CMD_SELECT);
+
+	/* we do not need to do anything if there are no CTEs */
+	if (queryTree->cteList == NIL)
+	{
+		return;
+	}
+
+	foreach(cteCell, queryTree->cteList)
+	{
+		CommonTableExpr *cte = (CommonTableExpr *) lfirst(cteCell);
+		Query *cteQuery = (Query *) cte->ctequery;
+
+		/*
+		 * Here we only check for command type of top level query. Normally there can be
+		 * nested CTE, however PostgreSQL dictates that data-modifying statements must
+		 * be at top level of CTE. Therefore it is OK to just check for top level.
+		 * Similarly, we do not need to check for subqueries.
+		 */
+		if (cteQuery->commandType != CMD_SELECT)
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot perform distributed planning for the given "
+								   "modification"),
+							errdetail("Only SELECT queries is supported as common table "
```

Comment:
PostgreSQL refers to these as _data-modifying statements in `WITH` clauses_. In fact, _common table expression_ is only mentioned once at the top of [the relevant documentation page](https://www.postgresql.org/docs/9.6/static/queries-with.html).

Let's use _Data-modifying statements are not supported in the WITH clauses of distributed queries_?